### ISSUE TITLE
Would be ideal to have some test case example

### ISSUE THREAD

**Issue by jimbasquiat** [erdewit/tws_async#3]
Hi Ewald,

Thanks a lot for sharing your implentation of the API with asyncio.

I'm quite new to asyncio, so it is a bit challenging to learn both about it and how to integrate it with the API.

Woult that be possible for you to post in the repo some simple test example of how to do a request (say for instance an historical data request) and another one to show how to do multiple requests?

Thank you,

Vincent

**Comment by jimbasquiat** [comment]
Hi Ewald, my question was obviously not clear but let me rephrase:

It seems to me that asyncio makes it possible to get rid of the nested callback architecture - which i think is difficult to read, maintain and expand.

But I am super fresh with that module so I thought i d rather ask to more experience people their opinion.

What I mean is that I expect it would be possible to go in the direction of such code, where the callback are defined first as part of the app class. Ideally that can be written has a module whom method I dont need to overload each time I work on the app to include the requests i need into various callback.
Ideally I could simply call the request i need in say, a "if __name__="__main__" at the bottom of the module or another script. Those request would need to have a way of being send sequentially, so they would need to know when the previous request has completed (use a function or create a flag that signal when the snapshotEnd, accountUpdateEnd, etc. callback has been received). 

I tried the following (obviously it doesnt work but to give an example of what i mean)

```
#!/home/depot/.conda/envs/python36/bin/python

from ibapi import wrapper
from ibapi.client import EClient
from ibapi.common import *
from ibapi.ticktype import *
from ibapi.contract import Contract
import datetime

import asyncio
from async import TWSClient

contract = Contract()
contract.symbol = "ES"
contract.secType = "FUT"
contract.exchange = "GLOBEX"
contract.currency = "USD"
contract.lastTradeDateOrContractMonth = "201709"

class TestApp(TWSClient):
    def __init__(self):
        TWSClient.__init__(self)
        self.end_request_time = datetime.datetime.now().strftime("%Y%m%d %H:%M:%S")
        
    @iswrapper
    def historicalData(self, reqId: TickerId, date: str, open: float, high: float,
                            low: float, close: float, volume: int, barCount: int,
                            WAP: float, hasGaps: int): 
        print("date", date, "open", open, "high", high,"low", low, "close", close, "volume", volume)
        
    @iswrapper
    def historicalDataEnd(self, reqId: int, start: str, end: str):
        print("HistoricalDataEnd ", reqId, "from", start, "to", end)
        self.done=True
 
if __name__ == "__main__":
    app = TestApp()
    app.connect("127.0.0.1", 4002, clientId=251)
    loop=asyncio.get_event_loop()
    loop.call_soon(app.reqHistoricalData(0, contract, app.end_request_time, "1 D", "10 mins", "TRADES", 0, 1, []))
    loop.run_forever()
    loop.close()
```



This gives an error as the reqhisto data is placed before the app has the time to connect. So indeed there would need to be a flag of some sort that could be used to wait for 1 call to be over to get to the next one.

I would expect the Future or Task function, or even the Call_soon could be used in such fashion?

This would be great.

**Comment by erdewit** [comment]
Closing this issue as it was discussed on the mailinglist (long time ago)
